Fix escape sequence parser for decoding strings

sfeuerstein-op opened the pull request:
Improved logic to parse escape sequences when decoding strings. The former gsub logic is replaced by a clean token parser.

This fixes the problem that the parser tries to decode any ``\u`` sequence (also invalid unicode escape sequences) when a valid escape sequence is found somewhere else in the string.

MNWE:

```lua
json = require("json")
json.decode("{\"fail\":\"\\\\url{http://www.example.com/} vs. \\u0023\"}")
```

rxi commented:
Good catch and thanks for submitting a pull request! A couple of problems:

- One of the tests in `test.lua` now fails
- The way you're building the string (`s = s .. string.char(x)`) in (at least non-JIT) lua is *slow*, allocating a new string, hashing and copying with each iteration. `table.concat` is provided for string building purposes

This is definitely something that needs fixing. An additional thought on the second point is how (with the change to `table.concat` string building) it would compare in performance on long strings to the old `:gsub` approach. Looking over the code now I feel it should be easy enough to fix the bug in the old approach -- if the `table.concat` approach does end up slower than the existing method I will probably do this instead.

Let me know if you're planning to continue working on this pull request, otherwise I'll probably fix the problem using a modified version of the existing method of handling escapes.

sfeuerstein-op commented:
Thanks for the criticism. I fixed the bug so that all tests are now passing again. I reworked string concatenation to use ``table.concat`` now. I also added some logic to not do a full concatenation when no active char was found. The current implementation should thus be a good bit faster than the the first iteration.

I did not do a benchmark of the implementation since ``bench_all.lua`` currently fails for me (seems like a missing dependency). I hope that this performs better or near to as good as the `gsub` approach since the logic should be way more robust than a pattern based approach.

sfeuerstein-op commented:
I also ran a benchmark of the two versions now. See the result below. Performance seems now to be nearly identical.

v0.1.2:
```
[decode]
Lua version   : Lua 5.1
CPU name      : Intel(R) Core(TM) i7-2600 CPU @ 3.40GHz
../json.lua   : 0.342s [x1] (min: 0.338s, max 0.348s)
dkjson.lua    : 0.347s [x1.01] (min: 0.343s, max 0.351s)
jfjson.lua    : 0.374s [x1.09] (min: 0.362s, max 0.389s)

[encode]
Lua version   : Lua 5.1
CPU name      : Intel(R) Core(TM) i7-2600 CPU @ 3.40GHz
../json.lua   : 0.148s [x1] (min: 0.146s, max 0.154s)
dkjson.lua    : 0.192s [x1.3] (min: 0.187s, max 0.203s)
jfjson.lua    : 0.256s [x1.74] (min: 0.248s, max 0.269s)
json4lua.lua  : 0.247s [x1.67] (min: 0.242s, max 0.255s)
```

master:
```
[decode]
Lua version   : Lua 5.1
CPU name      : Intel(R) Core(TM) i7-2600 CPU @ 3.40GHz
../json.lua   : 0.336s [x1] (min: 0.333s, max 0.339s)
dkjson.lua    : 0.35s [x1.04] (min: 0.347s, max 0.352s)
jfjson.lua    : 0.366s [x1.09] (min: 0.361s, max 0.371s)

[encode]
Lua version   : Lua 5.1
CPU name      : Intel(R) Core(TM) i7-2600 CPU @ 3.40GHz
../json.lua   : 0.147s [x1] (min: 0.145s, max 0.15s)
dkjson.lua    : 0.188s [x1.28] (min: 0.187s, max 0.192s)
jfjson.lua    : 0.256s [x1.74] (min: 0.251s, max 0.262s)
json4lua.lua  : 0.247s [x1.68] (min: 0.242s, max 0.253s)
```

rxi commented:
Thanks for the update and the benchmark results! Although the performance does seem comparable on puc lua, it unfortunately looks like it takes twice the time when ran on luajit:

```
Lua version   : LuaJIT 2.1.0-beta3
CPU name      : Intel(R) Core(TM) i5-4210U CPU @ 1.70GHz
../json.lua   : 0.0231s [x1.79] (min: 0.0216s, max 0.0256s)
old_json.lua  : 0.0129s [x1] (min: 0.0122s, max 0.0144s)
```
Interestingly enough I would have assumed this would have been the other way round, but the magic of JIT seems somewhat unpredictable.

I do favour your implementation and it does seem cleaner in general, and I appreciate the effort put into this pull request. I'm going to let it sit for the moment as I want to try modifying the old approach or explore other approaches as I'm not happy with the performance loss.

sfeuerstein-op commented:
I must say that i did not expect these result and that i did not benchmark my implementation with LuaJIT.

I did some more experiments with LuaJIT now and found that (while still slower than v0.1.2) the implementation in e1f3fdcd8834de07ac7e3108c7dc26d7a5089f9a is actually faster in LuaJIT than the current implementation. It might be an acceptable compromise.

Nevertheless, I can understand your concerns about the performance loss, yet I'm currently not sure how to further speed up the parser logic, so I might not be able to help any more here.

rxi commented:
I've pushed a commit which is similar to your approach, but without the loss in performance. I realised only afterwards that you had changed the code to use `string.find` with each iteration -- this might have been the cause of the slowness on the JIT. I thought it might be helpful to outline a few of the reasons for why the function was implemented the way it was:
* `string.byte` was used over `string.sub` as this is much faster on PUC lua -- `string.sub` would have to create a new interned string for each iteration, where as `string.byte` returns a number which on PUC lua are non-gced basic values. `string.sub` may have had no negative effect on JIT lua
* any pattern-matching functions were deferred and avoided if unneeded as these are not JIT compiled on JIT lua and are thus slow

Usually these kinds of things aren't a concern in Lua, but since this is a tight loop iterating on each byte of an input string which realistically could be large, the decision to have improved performance over readability was chosen.

Thanks again for your efforts and finding the bug in the first place!
